Lib/test/test_types.py

```diff
@@ -772,6 +772,39 @@ def test_union_parameter_chaining(self):
         self.assertEqual((list[T] | list[S])[int, T], list[int] | list[T])
         self.assertEqual((list[T] | list[S])[int, int], list[int])
 
+    def test_union_parameter_substitution(self):
+        def eq(actual, expected):
+            self.assertEqual(actual, expected)
+            self.assertIs(type(actual), type(expected))
+
+        T = typing.TypeVar('T')
+        S = typing.TypeVar('S')
+        P = typing.ParamSpec('P')
+        NT = typing.NewType('NT', str)
+        x = int | T | bytes
+
+        eq(x[str], int | str | bytes)
+        eq(x[list[int]], int | list[int] | bytes)
+        eq(x[typing.List], int | typing.List | bytes)
+        eq(x[typing.List[int]], int | typing.List[int] | bytes)
+        eq(x[typing.Hashable], int | typing.Hashable | bytes)
+        eq(x[collections.abc.Hashable],
+           int | collections.abc.Hashable | bytes)
+        eq(x[typing.Callable[[int], str]],
+           int | typing.Callable[[int], str] | bytes)
+        eq(x[collections.abc.Callable[[int], str]],
+           int | collections.abc.Callable[[int], str] | bytes)
+        eq(x[typing.Tuple[int, str]], int | typing.Tuple[int, str] | bytes)
+        eq(x[typing.Literal['none']], int | typing.Literal['none'] | bytes)
+        eq(x[str | list], int | str | list | bytes)
+        eq(x[typing.Union[str, list]], typing.Union[int, str, list, bytes])
+        eq(x[str | int], int | str | bytes)
+        eq(x[typing.Union[str, int]], typing.Union[int, str, bytes])
+        eq(x[NT], int | NT | bytes)
+        eq(x[S], int | S | bytes)
+        eq(x[P], int | P | bytes)
+        eq(x[typing.Concatenate[int, P]], int | typing.Concatenate[int, P] | bytes)
```

Comment:
`ParamSpec` and `Concatenate` cannot be unioned (or at least it is undefined behavior). So we don't have to test these. They don't throw `TypeError` because we decided to loosen type checks for them. In the future, these tests might break if their implementation changes. So it's best to leave them out.

Reply:
Then it is a separate bug.

Reply:
It's not really a bug, we made the decision quite early on to stop runtime type checking new types, and leave it to the type checker or other introspection libraries :). See Guido's comments https://bugs.python.org/msg382858.

Reply:
Jest verified -- no, defining `__or__` for `_TypeVarLike` is not a bug (it can make sense for `ParamSpec`), but in future we need to add additional checks for substitutions.

Reply:
Sorry, I wasn't very clear. Yes, ``__or__`` for ``_TypeVarLike`` is not a bug during runtime. But *using* it with union should mke no sense to the static type checker. Please see the grammar for the `ParamSpec` PEP:
https://www.python.org/dev/peps/pep-0612/#valid-use-locations